components/containers/partitioned_vector/include/hpx/components/containers/partitioned_vector/partitioned_vector_impl.hpp

```diff
@@ -102,10 +102,17 @@ namespace hpx
         typedef typename components::server::distributed_metadata_base<
             server::partitioned_vector_config_data>::get_action act;
 
+#if !defined(HPX_HAVE_CXX20_STD_LAMBDA_CAPTURE)
         return async(act(), id).then(
             [=](future<server::partitioned_vector_config_data>&& f) -> void {
```

Comment:
```suggestion
            [= HPX_CXX20_CAPTURE_THIS](future<server::partitioned_vector_config_data>&& f) -> void {
```
? Syntactically somewhat awkward but maybe nicer than the ifdefs? Alternatively I would place the if/else inside the capture list.

Reply:
@msimberg or else can i do this instead of if/else, i will use `HPX_CXX20_CAPTURE_THIS` as you suggested, but
define HPX_CXX20_CAPTURE_THIS as `, this` when `HPX_HAVE_CXX20_STD_LAMBDA_CAPTURE` is defined else to ` ` 
 empty space.

So instead of if/defs i just write  `HPX_CXX20_CAPTURE_THIS` in the captures as you shown above?

some thing like [this ](https://wandbox.org/nojs/gcc-head/permlink/ElKxIP2JLRaBM5tn).

Reply:
Yeah, that's exactly what I had in mind. @hkaiser any better ideas?

Reply:
I think having the ',' as part of the macro is not good. How about we use it as:
```
[HPX_CXX20_CAPTURE_THIS(=)](future<server::partitioned_vector_config_data>&& f) -> void {...}
```
where the argument list could be potentially empty. I.e.
```
#if HPX_HAVE_CXX20_STD_LAMBDA_CAPTURE
#define HPX_CXX20_CAPTURE_THIS(...) __VA_ARGS__, this
#else
#define HPX_CXX20_CAPTURE_THIS(...) __VA_ARGS__
#endif
```

Reply:
should it be like this?
```
HPX_PP_IF(HPX_PP_NARGS(__VA_ARGS__) == 0, \
        this, HPX_PP_IDENTITY(__VA_ARGS__, this))
```
assuming that `HPX_PP_IF` works like ternary `HPX_PP_IF(CONDITION, TRUE, FALSE)`

And `HPX_PP_IDENTITY` is bit simpler to implement, as
`#define HPX_PP_IDENTITY __VA_ARGS__`

but  any clue how `HPX_PP_IF` would be implemented?
because we cannot do something like this `#define SOME_MACRO #if ...`

Reply:
Shouldn't that be `HPX_PP_IF(HPX_PP_NARGS(__VA_ARGS__) == 0, this, HPX_PP_IDENTITY(__VA_ARGS__, this))`? And in the else branch `[HPX_CXX20_CAPTURE_THIS()]` would not be right. I have the feeling any approach is going to have inputs that are just plain wrong, so I guess we'd just have to pick from the least bad one. The reason I like my suggestion is that it's anyway only meant to be used when there's a `=` capture and it would make it painfully obvious that `[HPX_CXX20_CAPTURE_THIS]` is not valid. But I admit it's far from perfect as well. In the end I don't have a particularly strong opinion about this.

Reply:
this `HPX_CXX20_CAPTURE_THIS(...)`,
expects at least one macro argument right? so whether `HPX_CXX20_CAPTURE_THIS` is either a macro function or plain macro they both would work only if there is more than one argument?

If yes, then the solution could be one of the following?

```
#define HPX_PP_IDENTITY(...) __VA_ARGS__

#if HPX_HAVE_CXX20_STD_LAMBDA_CAPTURE
#define HPX_CXX20_CAPTURE_THIS(...) \
    HPX_PP_IDENTITY(__VA_ARGS__, this)
#else
#define HPX_CXX20_CAPTURE_THIS(...) __VA_ARGS__
#endif
```

or

`#define HPX_CXX20_CAPTURE_THIS , this`

Reply:
Yes, your suggestion of using identity is even simpler, nice. I would like to avoid having the `','` as part of the macro as this leads to 'invalidly' looking syntax: `[= HPX_CXX20_CAPTURE_THIS]`.

Reply:
> Shouldn't that be `HPX_PP_IF(HPX_PP_NARGS(__VA_ARGS__) == 0, this, HPX_PP_IDENTITY(__VA_ARGS__, this))`?

Yes, my bad, that's what I meant.

> And in the else branch `[HPX_CXX20_CAPTURE_THIS()]` would not be right.

Yes to this as well, However, I would expect for this to be used as `[HPX_CXX20_CAPTURE_THIS(=)]` only (with possibly more arguments).

Reply:
@hkaiser  thanks for this!